### DISCUSSION TITLE
How to access parent input argument from child data fetcher

### DISCUSSION THREAD

**Discussion by driptaroop** [Netflix/dgs-framework#23] [Q&A] [answered]
I have a graphQL schema,
```graphql
type Query {
    getContact (customerId: ID!): Contact
}

type Contact {
    email: String
    address: [Address]
}

type Address {
    street: String!
    number: String!
    additional: String
    pincode: String!
    city: String!
    country: String!
}
```

I generated the data models using the codegen plugin and wrote 2 data fetchers (contact and address). In the contact data fetcher, I can access the `customerId` using `@InputArgument` annotation. I am not able to access the `customerId` in the address data fetcher. I tried using `@InputArgument` (which getting null). Also, I tried `DataFetchingEnvironment.arguments` which is also resolving as an empty map.

code:
```kotlin
@DgsComponent
class ContactFetcher(
    private val addressService: AddressService
) {
    @DgsData(parentType = DgsConstants.QUERY_TYPE, field = DgsConstants.QUERY.GetContact)
    fun contact(@InputArgument("customerId") customerId: String): Contact = 
        Contact(email = "abc@abc.com") // here I can access the customerId

    @DgsData(parentType = DgsConstants.CONTACT.TYPE_NAME, field = DgsConstants.CONTACT.Address)
    fun address(dfe: DataFetchingEnvironment): List<Address> {
       /* I need to fetch the address based on the customer Id. but how to access customerId here?
          I tried dfe.arguments which is an empty map. @InputArgument does not work also.*/
        return addressService.getAddresses(customerId) 
    }
}

```

**Comment by paulbakker** [comment]
What you're likely looking for is the `getSource` method on `DataFetchingEnvironment`. This would give you the `Contact` in the address fetcher.

**Reply by driptaroop** [reply]
Hey @paulbakker , thanks for getting back to me. In this case, I can get the `Contact` from `dfe. getSource()` method, but only getting the `Contact` will not work for me. Because in order to fetch the address I need the `CustomerId` which is not a part of `Contact`.
As of now, the only way I can get `CustomerId`  in the `address` data fetcher is via passing it down through the `LocalContext` ([as described in this discussion](https://github.com/Netflix/dgs-framework/discussions/21#discussioncomment-343094)). Is there any other way to do this? or is this the recommended way?

**Reply by paulbakker** [reply]
I added another reply to that thread yesterday which is seems to be the answer to exactly what you're asking: https://github.com/Netflix/dgs-framework/discussions/21#discussioncomment-343440

Make sure to read the reply after the one linked as well, I've added a link to an example there.

**Reply by srinivasankavitha** [reply]
The input argument to the query is also available in the DataFetchingEnvironment's operationDefinition field that you can access like this: 
```code
// This gives you the first field that was selected on the query
Field result = (Field)(dfe.getOperationDefinition().getSelectionSet().getSelections().get(0));
// This gives you the first argument's value and so on
result.getArguments().get(0).getValue();
```

Admittedly, this is jumping through some hoops to get the value, so ultimately depends on what is convenient for you.

**Reply by driptaroop** [reply]
Thank you @paulbakker . I replied in the other thread.

Thank you @srinivasankavitha as well. I indeed got the value by this way although this is a very verbose way of getting it. A more intuitive way of getting the root arguments would make the usage a bit easier. I am thinking some annotation based method parameter injection (something like `@RootArguments`, working similarly as `@InputArgument` but for root arguments).  Is there something like this on the feature roadmap for DGS since this might be a common use case?

**Reply by yamalameyooooo** [reply]
```
// This gives you the first field that was selected on the query
Field result = (Field)(dfe.getOperationDefinition().getSelectionSet().getSelections().get(0));
// This gives you the first argument's value and so on
result.getArguments().get(0).getValue();
```

@srinivasankavitha i have got the similar use case, but in my case there are 4 argument and they can be given in any order, so managing them in code is tedious, can we have the same suggestion to get the values by argument names instead of getting it from list by index? as i may never know in 0th place what value for which argument is present

**Reply by srinivasankavitha** [reply]
That's hard to determine since this is basically using the API exposed by `graphql-java`. I would recommend having a deterministic order for the arguments that you can use to extract.

**Comment by srinivasankavitha** [comment]
That is a great suggestion since this is a common use case. We will add
this to the roadmap.

**Comment by srinivasankavitha** [comment, marked as answer]
Created a new issue for this - https://github.com/Netflix/dgs-framework/issues/36

**Reply by driptaroop** [reply]
Great! I will resolve this discussion. 👍🏼
